File: GrokkingDP/5/EditDistance.cpp
#include <climits>
#include <iostream>
#include <vector>

enum class Method { BRUTE, MEM, BU };

int EditDistance(std::string str1, std::string str2, Method m);
int Brute(std::string str1, std::string str2, size_t i1, size_t i2);
int Memoize(std::string str1, std::string str2, size_t i1, size_t i2,
            std::vector<std::vector<int>>& dp);
int BU(std::string str1, std::string str2);

int main(void) {
  Method m = Method::BRUTE;
  std::cout << EditDistance("bat", "but", m) << ":"
            << EditDistance("abdca", "cbda", m) << ":"
            << EditDistance("passpot", "ppsspqrt", m) << ":"
            << EditDistance("sdijfnweijf", "sdjifnwes", m);
  return 0;
}

int EditDistance(std::string str1, std::string str2, Method m) {
  std::vector<std::vector<int>> dp(str1.length(),
                                   std::vector<int>(str2.length()));
  switch (m) {
    case Method::BRUTE:
      return Brute(str1, str2, 0, 0);
    case Method::MEM:
      return Memoize(str1, str2, 0, 0, dp);
    case Method::BU:
      return BU(str1, str2);
  }
}

int Brute(std::string str1, std::string str2, size_t i1, size_t i2) {
  if (i1 == str1.length() || i2 == str2.length())
    return std::max(str1.length() - i1, str2.length() - i2);
  if (str1[i1] == str2[i2]) return Brute(str1, str2, i1 + 1, i2 + 1);
  int c1 = 1 + Brute(str1, str2, i1 + 1, i2 + 1);
  int c2 = 1 + Brute(str1, str2, i1 + 1, i2);
  int c3 = 1 + Brute(str1, str2, i1, i2 + 1);
  return std::min(c1, std::min(c2, c3));
}

int Memoize(std::string str1, std::string str2, size_t i1, size_t i2,
            std::vector<std::vector<int>>& dp) {
  if (i1 == str1.length() || i2 == str2.length())
    return std::max(str1.length() - i1, str2.length() - i2);
  if (dp[i1][i2] == 0) {
    if (str1[i1] == str2[i2]) {
      dp[i1][i2] = Memoize(str1, str2, i1 + 1, i2 + 1, dp);
    } else {
      int c1 = 1 + Memoize(str1, str2, i1 + 1, i2 + 1, dp);
      int c2 = 1 + Memoize(str1, str2, i1 + 1, i2, dp);
      int c3 = 1 + Memoize(str1, str2, i1, i2 + 1, dp);
      dp[i1][i2] = std::min(c1, std::min(c2, c3));
    }
  }
  return dp[i1][i2];
}

int BU(std::string str1, std::string str2) {
  int len1 = str1.length();
  int len2 = str2.length();
  std::vector<std::vector<int>> dp(len1 + 1, std::vector<int>(len2 + 1));
  for (int i = 1; i <= len1; ++i)
    dp[i][0] = i;
  for (int j = 1; j <= len2; ++j)
    dp[0][j] = j;
  for (int i = 1; i <= len1; ++i) {
    for (int j = 1; j <= len2; ++j) {
      if (str1[i - 1] == str2[j - 1])
        dp[i][j] = dp[i - 1][j - 1];
      else
        dp[i][j] = 1 + std::min(dp[i - 1][j - 1], std::min(dp[i - 1][j], dp[i][j - 1]));
    }
  }
  return dp[len1][len2];
}

File: GrokkingDP/5/edit_distance.py
#  Problem Statement #
#  Given strings s1 and s2, we need to transform s1 into s2 by deleting, inserting, or replacing characters. Write a function to calculate the count of the minimum number of edit operations.

#  Example 1:

#  Input: s1 = "bat"
#  s2 = "but"
#  Output: 1
#  Explanation: We just need to replace 'a' with 'u' to transform s1 to s2.

#  Example 2:

#  Input: s1 = "abdca"
#  s2 = "cbda"
#  Output: 2
#  Explanation: We can replace first 'a' with 'c' and delete second 'c'.

#  Example 3:

#  Input: s1 = "passpot"
#  s2 = "ppsspqrt"
#  Output: 3
#  Explanation: Replace 'a' with 'p', 'o' with 'q', and insert 'r'.

t1 = ('bat', 'but')
t2 = ('abdca', 'cbda')
t3 = ('passpot', 'ppsspqrt')

def td(s1: str, s2: str) -> int:
    """
    >>> td(*t1)
    1
    >>> td(*t2)
    2
    >>> td(*t3)
    3
    """
    A, B = len(s1), len(s2)
    dp = [[None]*B for _ in range(A)]
    def fn(i: int, j: int) -> int:
        if i == A and j == B: return 0
        if i == A or j == B: return float('inf')
        if dp[i][j] is None:
            if s1[i] == s2[j]: dp[i][j] = fn(i+1, j+1)
            else: dp[i][j] = 1 + min(fn(i+1, j+1), fn(i+1, j), fn(i, j+1))
        return dp[i][j]
    return fn(0, 0)

def bu(s1: str, s2: str) -> int:
    """
    >>> bu(*t1)
    1
    >>> bu(*t2)
    2
    >>> bu(*t3)
    3
    """
    A, B = len(s1), len(s2)
    if B > A: A, B, s1, s2 = B, A, s2, s1
    dp, cnt = [[0]*B for _ in range(2)], 0
    yv = xv = False
    for a in range(B):
        if s2[a] == s1[0] and not xv: xv = True
        else: cnt += 1
        dp[0][a] = cnt
    for i in range(1, A):
        for j in range(B):
            if j == 0:
                if s1[i] == s2[j] and not yv: yv = True
                else: dp[1][j] = dp[0][j] + 1
            else:
                if s1[i] == s2[j]: dp[1][j] = dp[0][j-1]
                else: dp[1][j] = 1 + min(dp[0][j], dp[1][j-1], dp[0][j-1])
        dp[0] = dp[1][:]
    return dp[1][-1]

if __name__ == '__main__':
    import doctest
    doctest.testmod()
